cparser/Elab.ml

```diff
@@ -1523,49 +1567,49 @@ let elab_expr vararg loc env a =
   | TYPE_SIZEOF (spec, dcl) ->
       let (ty, env') = elab_type loc env spec dcl in
       if wrap incomplete_type loc env' ty then
-        error "incomplete type %a" Cprint.typ ty;
+        error "invalid application of 'sizeof' to an incomplete type %a" (print_typ env) ty;
       { edesc = ESizeof ty; etyp = TInt(size_t_ikind(), []) },env'
 
   | EXPR_ALIGNOF a1 ->
       let b1,env = elab env a1 in
       if wrap incomplete_type loc env b1.etyp then
-        error "incomplete type %a" Cprint.typ b1.etyp;
+        error "invalid application of 'alignof' to an incomplete type %a" (print_typ env) b1.etyp;
       { edesc = EAlignof b1.etyp; etyp =  TInt(size_t_ikind(), []) },env
 
   | TYPE_ALIGNOF (spec, dcl) ->
       let (ty, env') = elab_type loc env spec dcl in
       if wrap incomplete_type loc env' ty then
-        error "incomplete type %a" Cprint.typ ty;
-      { edesc = EAlignof ty; etyp =  TInt(size_t_ikind(), []) },env
+        error "invalid application of 'alignof' to an incomplete type %a" (print_typ env) ty;
+      { edesc = EAlignof ty; etyp =  TInt(size_t_ikind(), []) },env'
 
   | UNARY(PLUS, a1) ->
       let b1,env = elab env a1 in
       if not (is_arith_type env b1.etyp) then
-        err "argument of unary '+' is not an arithmetic type";
+        error "invalid argument type %a to unary expression" (print_typ env) b1.etyp;
       { edesc = EUnop(Oplus, b1); etyp = unary_conversion env b1.etyp },env
 
   | UNARY(MINUS, a1) ->
       let b1,env = elab env a1 in
       if not (is_arith_type env b1.etyp) then
-        err "argument of unary '-' is not an arithmetic type";
+        error "invalid argument type %a to unary expression" (print_typ env) b1.etyp;
       { edesc = EUnop(Ominus, b1); etyp = unary_conversion env b1.etyp },env
 
   | UNARY(BNOT, a1) ->
       let b1,env = elab env a1 in
       if not (is_integer_type env b1.etyp) then
-        err "argument of '~' is not an integer type";
+        error "invalid argument type %a to unary expression" (print_typ env) b1.etyp;
       { edesc = EUnop(Onot, b1); etyp = unary_conversion env b1.etyp },env
 
   | UNARY(NOT, a1) ->
       let b1,env = elab env a1 in
       if not (is_scalar_type env b1.etyp) then
-        err "argument of '!' is not a scalar type";
+        error "invalid argument type %a to unary expression" (print_typ env) b1.etyp;
       { edesc = EUnop(Olognot, b1); etyp = TInt(IInt, []) },env
 
   | UNARY(ADDROF, a1) ->
       let b1,env = elab env a1 in
       if not (is_lvalue b1 || is_function_type env b1.etyp) then
-        err "argument of '&' is not an l-value";
+        err "cannot take the address of an rvalue of type %a" (print_typ env) b1.etyp;
```

Comment:
The original message is better.  "rvalue" is not ISO C terminology.  What matters here is not that we have a rvalue, but that we have something that is not a lvalue!